art/attacks/evasion/graphite/graphite_whitebox_pytorch.py

```diff
@@ -0,0 +1,578 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the white-box version of the GRAPHITE attack `GRAPHITEWhiteboxPyTorch`.
+This is a robust physical perturbation attack.
+
+| Paper link: https://arxiv.org/abs/2002.07088
+| Original github link: https://github.com/ryan-feng/GRAPHITE
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Optional, Tuple, Union, TYPE_CHECKING, List
+
+import numpy as np
+
+from art.config import ART_NUMPY_DTYPE
+from art.attacks.attack import EvasionAttack
+from art.estimators.estimator import BaseEstimator
+from art.estimators.classification import ClassifierMixin
+from art.utils import compute_success, to_categorical, check_and_transform_label_format, is_probability
+from art.attacks.evasion.graphite.utils import convert_to_network, get_transform_params, transform_wb
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+    import torch
+
+logger = logging.getLogger(__name__)
+
+
+class GRAPHITEWhiteboxPyTorch(EvasionAttack):
+    """
+    Implementation of the white-box GRAPHITE attack from Feng et al. (2022). This is a physical attack
+    that generates robust physical perturbations that can be applied as stickers.
+
+    | Paper link: https://arxiv.org/abs/2002.07088
+    | Original github link: https://github.com/ryan-feng/GRAPHITE
+    """
+
+    attack_params = EvasionAttack.attack_params + [
+        "net_size",
+        "min_tr",
+        "num_xforms",
+        "step_size",
+        "steps",
+        "first_steps",
+        "patch_removal_size",
+        "patch_removal_interval",
+        "num_patches_to_remove",
+        "rand_start_epsilon_range",
+        "rotation_range",
+        "dist_range",
+        "gamma_range",
+        "crop_percent_range",
+        "off_x_range",
+        "off_y_range",
+        "blur_kernels",
+        "batch_size",
+    ]
+    _estimator_requirements = (BaseEstimator, ClassifierMixin)
+
+    def __init__(
+        self,
+        classifier: "CLASSIFIER_NEURALNETWORK_TYPE",
```

Comment:
This seems to only support PyTorchClassifiers based on the inputs to the other methods?